system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -0,0 +1,305 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jstring.hpp"
+#include "jdebug.hpp"
+#include "jptree.hpp"
+#include "jexcept.hpp"
+#include "jlog.hpp"
+
+#include <libxml/xmlmemory.h>
+#include <libxml/parserInternals.h>
+#include <libxml/debugXML.h>
+#include <libxml/HTMLtree.h>
+#include <libxml/xmlIO.h>
+#include <libxml/xinclude.h>
+#include <libxml/catalog.h>
+#include <libxml/xpathInternals.h>
+#include <libxml/xpath.h>
+#include <libxml/xmlschemas.h>
+#include <libxml/hash.h>
+
+#include "xpathprocessor.hpp"
+
+class CLibCompiledXpath : public CInterface, public ICompiledXpath
+{
+private:
+    xmlXPathCompExprPtr m_compiledXpathExpression = nullptr;
+    StringBuffer m_xpath;
+    ReadWriteLock m_rwlock;
+
+public:
+    IMPLEMENT_IINTERFACE;
+
+    CLibCompiledXpath(const char * xpath)
+    {
+        m_xpath.set(xpath);
+        m_compiledXpathExpression = xmlXPathCompile(BAD_CAST m_xpath.str());
+    }
+    ~CLibCompiledXpath()
+    {
+        xmlXPathFreeCompExpr(m_compiledXpathExpression);
+    }
+    const char * getXpath()
+    {
+        return m_xpath.str();
+    }
+    xmlXPathCompExprPtr getCompiledXPathExpression()
+    {
+        return m_compiledXpathExpression;
+    }
+};
+
+class CLibXpathContext : public CInterface, public IXpathContext
+{
+private:
+    xmlDocPtr m_xmlDoc = nullptr;
```

Comment:
The xpath context is shared by all threads of the same binding, so the XmlDocPtr can not be a member variable. One thread sets it, another thread may set it to a different doc, while the first thread is still using it. You should pass this in as a parameter to a method, or create another class to contain it that will have an object instantiated for every request/thread.